x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/blob/BlobStoreCacheMaintenanceService.java

```diff
@@ -0,0 +1,229 @@
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.index.reindex.BulkByScrollResponse;
+import org.elasticsearch.index.reindex.DeleteByQueryAction;
+import org.elasticsearch.index.reindex.DeleteByQueryRequest;
+import org.elasticsearch.repositories.IndexId;
+import org.elasticsearch.snapshots.SearchableSnapshotsSettings;
+import org.elasticsearch.snapshots.SnapshotId;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Supplier;
+
+import static org.elasticsearch.gateway.GatewayService.STATE_NOT_RECOVERED_BLOCK;
+import static org.elasticsearch.xpack.core.ClientHelper.SEARCHABLE_SNAPSHOTS_ORIGIN;
+import static org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshots.SNAPSHOT_INDEX_ID_SETTING;
+import static org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshots.SNAPSHOT_INDEX_NAME_SETTING;
+import static org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshots.SNAPSHOT_SNAPSHOT_ID_SETTING;
+import static org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshots.SNAPSHOT_SNAPSHOT_NAME_SETTING;
+
+/**
+ * A service that delete documents in the snapshot blob cache index when they are not required anymore.
+ *
+ * This service runs on the data node that contains the snapshot blob cache primary shard. It listens to cluster state updates to find
+ * searchable snapshot indices that are deleted and checks if the index snapshot is still used by other searchable snapshot indices. If the
+ * index snapshot is not used anymore then i triggers the deletion of corresponding cached blobs in the snapshot blob cache index using a
+ * delete-by-query.
+ */
+public class BlobStoreCacheMaintenanceService implements ClusterStateListener {
+
+    private static final Logger logger = LogManager.getLogger(BlobStoreCacheMaintenanceService.class);
+
+    private final ClusterService clusterService;
+    private final Client clientWithOrigin;
+    private final String systemIndexName;
+    private final ThreadPool threadPool;
+
+    public BlobStoreCacheMaintenanceService(ClusterService clusterService, ThreadPool threadPool, Client client, String systemIndexName) {
+        this.clientWithOrigin = new OriginSettingClient(Objects.requireNonNull(client), SEARCHABLE_SNAPSHOTS_ORIGIN);
+        this.systemIndexName = Objects.requireNonNull(systemIndexName);
+        this.clusterService = Objects.requireNonNull(clusterService);
+        this.threadPool = Objects.requireNonNull(threadPool);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        final ClusterState state = event.state();
+        if (state.getBlocks().hasGlobalBlock(STATE_NOT_RECOVERED_BLOCK)) {
+            return; // state not fully recovered
+        }
+        if (event.indicesDeleted() == null || event.indicesDeleted().isEmpty()) {
```

Comment:
I am not entirely sure this catches deletes that happen while the shard is red, i.e., all nodes containing the shard are offline?

Reply:
I thought a bit more about this and I think that you are right, it's possible that some indices deletions got bypassed if the `.snapshot-blob-cache` is unassigned for whatever reason.

A more solid solution would be to run this service on the master node and always try to delete cache entries whatever the status of the `.snapshot-blob-cache` cache is. Then we could complete this by running a full cache clean up every day that would list the entries and remove the ones that do not belong to mounted indices. What do you think?

Reply:
I think we could still keep it on the 0'th primary shard node. But running a cleanup like you describe once per day could help catch any misses. I think that should work out fine too.

Reply:
Thanks Henning. I'll update this PR accordingly and I'll add a daily task in a follow up PR.

Reply:
I opened #78438 for the full clean up.